README.md
# **Arduino Race Timer and Lap Gate Controller**
Download this repository into a local folder with the same name as the contained `.ino` to use as an Arduino sketch.

### [This video](https://youtu.be/n0GLdoQ6pg4) provides a live use demonstration of the example implementation hardware.
NOTE: the video demo is of an earlier version of the system. With the release of V2.X, many new features have been added that are not demonstrated in the video, including start lights, drag racing mode, a race pre-stage step, and other user customizations of race parameters.


---
## Table of Contents
1. [Introduction](#introduction)
    - [Prerequisites](#prerequisites) 
2. [Hardware Configuration](#hardware-configuration)
    - [Pin Out for Arduino Nano](#pinout-for-wiring-arduino-nano)
    - [Wiring Diagram](#wiring-diagram)
    - [Adjustments for Using ATMega2560 Based Arduino Instead](#changes-if-using-atmega2560-microcontroller-based-module)
3. [Software Configuration](#software-configuration)
4. [The Main Display (LCD2004 + I2C Backpack)](#main-display)
    - [Creating Custom Characters](#custom-characters)
5. [Racer Lap Timers (8-digit, 7-seg LED Bar)](#lap-timers)
    - [ Customization of LedControl library ](#customization-of-the-ledcontrol-librarys-character-table) - to display certain missing characters on 7-seg LED digits, replace `LedControl` library character table with an [updated table](#updated-chartable-to-replace-the-one-found-in-ledcontrolh).
6. [DIY MAX7219 Start Light Tree](#diy-max7219-start-light-tree)
7. [Adafruit Bi-Color Bargraph-24 as a Start Light](#adafruit-bi-color-bargraph-24-as-a-start-light)
8. [The Keypad](#the-keypad)
9.  [Lap/Gate Sensing](#gate-sensing)
    - [Drag Racing Sensor Setup](#drag-racing-sensor-setup)
10. [Sensor Options](#sensor-options)
11. [Analog Pause and Start Buttons](#analog-pause--start-buttons)
12. [Playing Audio](#playing-audio)
     - [Songs & Melodies](#songs--melodies)
     - [Method 1](#songs-method-one) - Using `Notes[]` and `Lengths[]` arrays
     - [Method 2](#songs-method-two) - Using RTTTL Ringtones
13. [Audio Modes](#audio-modes)
14. [Race Controller Operation](#race-controller-operation)
    - [Main Menu Navigation](#main-menu-navigation)
    - [Racing](#racing)
15. [Customizing UI Text and General Controller Settings](#customizing-ui-text-and-general-controller-settings)

<br>

# **Introduction**
This is an Arduino based project that implements an inexpensive, full featured, race game controller that can be used for timed racing games of 1-4 racers. Though the primary application is for slot car racing, the implementation is generally suited for any lane/gate based application. A [local configuration file](#config-file) can be maintained, to customize menus, sounds, default setup, and many other system parameteres.

**Key Features** - The primary features of the system include:
- 1-4 racer lanes/gates
- 3 racing modes (Total Laps & Timed Circuit Racing + Drag Racing)
- 0.000 ms timing precision
- A 4x20Ch LCD Primary Display
- 16 key, keypad interface
- Support for an 8 digit LED lap timer display for each lane/racer
- 3 audio modes (All, GameOnly, Mute)
- Easy system customization and localization via local config file
- Start light tree support (DIY MAX7219 Tree and/or Adafruit Bargraph-24)

In the presented configuration, the lap sensing input is simulated using buttons, but can be adapted to be used with a myriad of simple, circuit completion, or other type sensing methods that can be implemented in the physical lap gate. The working demo of this project uses two paper clips integrated into a mechanical lap counter to create a simple, yet effective lap sensor.

See [Race Controller Operation](#race-controller-operation) at the end of this document for menu navigation and general use.

![MainMenu](Images/ScreenShots/Main_Menu.png)  
![Wiring Diagram](Images/FullSystem.png)  
![Wiring Diagram](Images/Breadboard_4Racer.png)  



## **Prerequisites**  <a id="prerequisites"></a>
This project involves a lot of different hardware and software concepts, and implements some more intermediate to advanced code for Arduino.

However, I have endeavored to explain in enough detail that someone with almost no experience can still follow along, and successfully build a working controller. It will even be possible to customize menu text and default game settings, without having to write new code.

It is expected the reader understands how to use the Arduino IDE, connect wires, and program boards. To get up to speed on those basics, please check out, [Getting Started With Arduino Projects](https://www.arduino.cc/en/Guide), or many other great resources around the web.

> ***Code Snippets in Readme** - the code snippets shown in this readme file are not meant to be cut and paste working examples. They are used only to illustrate the syntax for general setup and usage. For working code, reference the releated sections of the actual project code in `RaceTimerAndController.ino`*  

> ***Note on Reference Sources** - All links are for reference only and are not to be taken as an endorsement of any particular component supplier. I attempt to reference official Arduino resources whenever possible, but this is also not an endorsement for or against using the Arduino store.*


<br>

# **Hardware Configuration**  
All of the components are readily available and can be connected with basic jumper leads or simple conductors (wires). Some components might require header pin soldering if they do not come pre-assembled.
> ***Note on Housing and Mechanical Interface** - This project only documents the functional electrical and software configuration. It can be wired, and used as illustrated for demonstration, however, for repeated, practical usage, the construction of a permanent housing, and mechanical trigger interface, is left up to the implementer to develop per their unique setup.*  

## **Parts for Core Race Controller**
The parts required for the base system are listed below. The 7-segment, racer lap timer displays, will update with the same indications as a start light tree, so all race modes can be run without a start light, however, optionally, two types of start lights are also supported.
- [Arduino Nano](https://docs.arduino.cc/hardware/nano) (or equivalent ATMega328 based microcontroller module) ([amazon search](https://www.amazon.com/arduino-nano/s?k=arduino+nano))
  - if using Mega2560 adjust code and wiring per annotations)
- [4 x 4 membrane keypad](https://duckduckgo.com/?q=4+x+4+membrane+keypad)
- [LCD2004 4 row x 20 character display](https://duckduckgo.com/?q=LCD2004A+4+x+20+I2C+backpack), with [I2C backpack](https://www.mantech.co.za/datasheets/products/LCD2004-i2c.pdf)
- 2-4 Chainable, [8-digit, 7-segment LED bar with integrated MAX7219](https://duckduckgo.com/?q=8-digit%2C+7-segment+LED+display)
- Passive Buzzer or speaker
  - The 7-seg LEDs induced a hum on my buzzer, I used a diode on one lead to eliminate it.
- 2-4 Analog or digital lap sensors/switches/buttons, 1 for each lane.
    - (optional) +2 drag race finish sensors - Drag racing can be setup to use just a finish sensor, however, to support the default start & finish drag sensing, add 2 finish sensors that will share the same pins as the start sensors for lanes 1 & 2.
- 2 momentary switches for a 'Pause' and optional 'Start' button.
  - 10k Ohm Pull-Up resistor for each analog input (button) on `A6` and/or `A7`.
- Jumper leads to wire connections between peripherals & Arduino

## Supported Start Light Configurations
Starting with Ver2.0, this controller will support using an Adafruit Bi-color Bargraph-24, and/or, a custom built LED light tree that can be assembled using standard LED components and a MAX7219 serial dirver.

### DIY MAX7219 Light Tree Parts
These are all the parts required to build a custom start light tree. For further details, see section: [DIY MAX7219 Based Start Light Tree](#diy-max7219-based-start-light-tree).
- [MAX7219 Serial LED driver chip](https://www.amazon.com/Bridgold-MAX7219CNG-MAX7219-Serially-Interfaced/dp/B0BK6XC4P8/ref=sr_1_3?keywords=max7219&sr=8-3) - ([datasheet](https://www.analog.com/media/en/technical-documentation/data-sheets/MAX7219-MAX7221.pdf)) same chip used by the pre-packaged LED bars.
- 4 White LEDs
- 6 Yellow LEDs
- 2 Green LEDs
- 2 Red LEDs
- 2 Blue LEDs
- 100nF (i.e. 0.1μF, #104) capacitor
- 10k Ohm Resistor

### Adafruit Start Light Parts
- [Adafruit Bi-Color Bargraph-24, With I2C backpack kit](https://www.adafruit.com/product/1721) - This is an I2C compatible, full assembly of an LED light bar with 24 segments. Each segment has a Green, and a Red, LED, which can be lit to create red, yellow, or green.


## **PinOut for Wiring Arduino Nano**
(these are the pins used by this code, but can be re-arranged, if desired)  
<div style="color:orange;font-size:12px">  

|Attached        |Pin  |Nano|Pin  |Attached        |
|---------------:|----:|:--:|-----|----------------|
|Passive Buzzer  |D13  |\-- |D12  |Keypad 8 (C4)   |
|                |3V3  |\-- |D11  |Keypad 7 (C3)   |
|                |Free |\-- |D10  |Keypad 6 (C2)   |
|Lane1 lap sensor|A0   |\-- |D9   |Keypad 5 (C1)   |
|Lane2 lap sensor|A1   |\-- |D8   |Keypad 4 (R4)   |
|Lane3 lap sensor|A2   |\-- |D7   |Keypad 3 (R3)   |
|Lane4 lap sensor|A3   |\-- |D6   |Keypad 2 (R2)   |
|LCD/Bargraph SDA|A4   |\-- |D5   |Keypad 1 (R1)   |
|LCD/Bargraph SCL|A5   |\-- |D4   |LED CLK         |
|Pause Button    |A6   |\-- |D3   |LED CS          |
|Start Button    |A7   |\-- |D2   |LED DIN         |
|                |5V   |\-- |GND  |common GND      |
|                |Reset|\-- |Reset|                |
|common GND      |GND  |\-- |Rx0  |                |
|+5V Supply      |Vin  |\-- |Tx1  |                |
</div>  

## **Wiring Diagram**
![Wiring Diagram](Images/V2X_Wiring_Diagram1600x1600.png)

## **Changes if Using ATMega2560 Microcontroller Based Module**
By default the code is setup for an Arduino Nano, ATMega328 based pin out. However, with minor adjustments, this project can also work with modules, like the Arduino Mega2560, that use an ATmega2560 processor. To switch between using a ATMega328 and ATMega2560, one can edit the hardware variables referenced in the `localSettings.h` file.

The summary of the changes to be made to use a Mega2560 board are:
- Use pins `D20` & `D21` for LCD's SDA and SCL (respectively), instead of `A4` & `A5`.
- For the lane sensors, use pins on I/O `port K` instead of `port C`.
  - Wire lanes1-4 to pins A8-A11 instead of pins A0-A3
  - Edit setting `PCINT_VECT` in the `localSettings.h` file, to change interrupt vector used to be `PCINT2_vect`, instead of `PCINT1_vect`.
  - and accordingly, edit setting `INTERRUPT_PORT` in the `localSettings.h` file, to read from the port K byte, `PINK`, instead of port C byte, `PINC`, to check triggered lanes.

## **Power Supply (+5V)**  
All devices in this build are powered from a +5V source. The displays should draw power from the source supply and not through the Arduino which cannot support enough current to run everything properly.

> ***Powering MAX7219 LED Bars** - Power for these can be daisy chained for the first 2 bars, but cascading 3 or more may require running the power directly to each subsequent display bar, but always keep all the signal lines daisy chained.*

> ***<span style="color:yellow"> Connect Arduino GND to external ground reference </span>** - Like many projects with higher power demands, this one uses an external power supply to get enough current for operation; during development and programming, we will often have the computer USB plugged in as well, which also serves as a power supply. Sometimes the computer might be using battery and sometimes it may be using wall power, this can change the ground reference of the USB with respect to the other power sources. If one is not careful, one may create a ground reference mismatch. To avoid this we should jumper all common input supply grounds into the Arduino. Ground reference mismatch can cause intermittent errors, or the device to not work at all.*  
> 
> *If we were to inadvertently find ourselves with the configuration shown here. It may appear to be ok on first look, but with the multi-meter we can see that there is a 1.9V differential between the two ground references when it should be reading close to 0.*
> 
> ![GroundLoop HIGH](Images/MismatchedGroundRef_HIGH.png)
> Connecting our grounds to bring them to the same potential, as below, will eliminate the problem above.  
> 
> ![Connected Grounds](Images/MismatchedGroundRef_LOW.png)
 
<br>

# **Software Configuration**  
In order to interact with our different peripherals, this project uses several existing [Arduino libraries](https://www.arduino.cc/reference/en/libraries/). Unless otherwise specified, these libraries can be downloaded the usual manner using the [Arduino Library Manager](https://docs.arduino.cc/software/ide-v1/tutorials/installing-libraries). Each library will be introduced with the hardware it's related to.

Reference for add-on libraries used in the core project. The versions listed are the latest, verified compatible versions. If newer versions are causing a compiling issue, try the version listed below:
- [hd44780](https://www.arduino.cc/reference/en/libraries/hd44780/) v1.3.2
- [LedControl](https://www.arduino.cc/reference/en/libraries/ledcontrol/) v1.0.6
- [Keypad](https://www.arduino.cc/reference/en/libraries/keypad/) v3.1.1
- [PlayRtttl](https://github.com/ArminJo/PlayRtttl) v1.4.2 (at the time of this edit, it is known version 2.x+ of this library are currently NOT compatible with this project, please make sure to install V1.4.2)
- ^ [Adafruit GFX](https://www.arduino.cc/reference/en/libraries/adafruit-gfx-library/) 1.11.5
- ^ [Adafruit LED Backpack](https://www.arduino.cc/reference/en/libraries/adafruit-led-backpack-library/) 1.3.2

^ These libraries are only necessary to support the Adafruit Bargraph. However they must still be installed, even if not using them, to avoid any compiling errors.

All custom project logic is in the main Arduino `.ino` sketch file. The additional supporting `.h` files are used to store different data constants defining custom characters, songs, and default system settings.
- `RaceTimerAndController.ino` - main sketch file containing all custom logic
- `enum_list.h` - all defined enums are put in a seperate file to force the compiler to handle them before they are used as a global function parameter. Not doing so can cause a compiler error, even though the code is ok.
- `RTTL_songs.h` - file to hold RTTL song string constants
- `melodies_prog.h` - file to hold Notes[] Lengths[] array based song data
- `pitches.h` - file holding `#define` macros setting the frequency values used for Notes array based songs.
- `CustomChars.h` - file holding byte array constants that define the shape of custom icons used in this project.
- `defaultSettings.h` - file holding `#define` macros that establish the values used for the default controller UI text and race settings.

CHANGE_LOG.md
# Change Log
This document records the changes beteween updates of the 'RaceTimerAndController" code repository.

The versioning scheme will conform to the following format:

Ver. ##.##.## - Major.Minor.Patch
- **Major version #** - Increments on changes that are not backward compatible
- **Minor version #** - Increments on changes that add/remove functionlity, but can be used, without issue, in place of earlier versions of the same major release
- **Patch version #** - Increments when update is a bug fix, refactoring, or new data constant (like a song & custom ch arrays), as well as on updates to non functional aspects of the code repo, like documentation.


## Ver 2.0.0 - Added Drag Racing & Start Button, Pause Button Debounce Bug Fix
> This update is a significant update to the code. Though technically, this code is backward compatible with the same hardware setup as version, 1.x, the code base infrastructure has significant changes, and internal variable consistency is broken from earlier versions.
> 
> **Updates from Previous Commit**
>
> New Features/Enhancements
> - **Drag Racing Support** - This Race Controller now supports Drag Racing on lanes 1 & 2, with a new 'Drag Race' option, selectable from the 'Start a Race' menu. Drag racing can be setup with a single set of finish sensors, or with 2 sets of sensors, 1 at start, and 1 at finish. With only finish sensors, racers must make there own call on false starts. By default, the controller will expect a start and a finish sensor in each lane. Sensor configuration can be changed by editing the `SINGLE_DRAG_TRIGGER`flag in `localSettings.h`.
> NOTE: The number of sensors does not change the pin assignments. Both the start and finish sensors for a given lane should be wired as open switches on the same pin.
> - **Race Pre-Stage** - Previously, the race pre-start countdown started immediately after selecting a race type from the **Start a Race Menu**. Now, a 'Pre-Stage' has been added alerting racers to get to starting positions. When all racers are ready, the race can then be started by pressing the '#' key or the new **Start Button**.
>     - This extra phase facilitates rapid restarts of heats in drag racing.
>     - Also, coupled with adding a **Start Button**, the Pre-Stage allows for a more flexible system setup. Being able to start the race from a button allows the starter and controller's main keypad to be in different locations.
> - **Adafruit Bargraph** - The system now supports using an [Adafruit 24 Bi-Color LED Bargraph](https://www.adafruit.com/product/1721) as a start light tree and general race state indicator. The bargraph uses I2C and should be wired, sharing the same pins as the main LCD display.
> - **MAX7219 LED Start Light Tree** - The system now supports using a DIY, MAX7219 based, custom LED start light. The start light should be wired as the last device in the racer LED bar chain. The system will assume that the number of 7-seg race timers will equal the number of racers defiend by `LANE_COUNT`. As such it will assume the start light will be at device index `LANE_COUNT`, since device indexing starts with 0. See the project documentation for how to build the start light.
> - **Start Button** - Support for a 2nd analog buttton has been added. This button can used to trigger a race start, or restart, as well as clear start faults. The hardware wiring, and setup, mirrors the analog **Pause Button**. By default, the **Start Button** uses `PIN_A7`, but that can be changed using the `localSettings.h` file.
> - **Start Fault** - Previously with only circuit racing modes, the lap sensing pins were disabled until the race started. Now, lap triggers are enabled at the initiaton of the **Pre-Start** stage and crossing the start before the end of the  **Pre-Start** countdown will trigger a fault in all race modes. The offending racer/lane will be displayed on the main LCD and the associated lane's fault indicator LED, on start light tree, will light. Pressing the **Pause** or **Start** button, or the `#` or `*` key, will clear the fault and return to the `Staging` state.
> - **Song Association** - Users can now use the `SONGS_BY_PLACE` flag in `localSettings.h` to specify if the `RACERS_SONGS_LIST` should be associated with a place finish or with a particular racer. By default, songs are associated with racers, however, by setting `SONGS_BY_PLACE` to `true`, the song at index 1 will play when the 1st place finishes, the song at index 2 will play when the 2nd racer finishes, and so on.
> - **Restart Countdown** - Users can now use the `CTDWN_ON_RESTART` flag in `localSettings.h` to specify whether restarts after a pause, should start right away or if it should restart with a pre-start countdown.
>
> Bug Fixes
> - MAJOR BUG FIX - Pause button debouncing was not properly filtering the debounce time because the time was being tracked by an `int` variable instead of an `usinged long` which caused an error in the debounce calculation, effectively nullifying it.
>
> Notable Code Changes
> - All enums have been moved to a seperate file and imported into the main .ino. This forces the compiler to look at those first, making them available to be used in function definitions without throwing a compiler error.
> - New states have been created, `PreStart`, `Staging`, `Fault`, `PreFault`, and `Finished`, to better manage system status flow.
> - The need for `topFastestLaps[]` and `topFastestTimes[]` has been eliminated and their function rolled into the, previously unused, zero index of the main `fastestLaps[0][0]` and `fastestTimes[0][0]` arrays used by the racers.
> - Moved LCD and LED pin-outs and parameters to be `...Settings.h` macros
________________________

## Ver 1.1.0 - Implementation of Adjustable Lane Count

defaultSettings.h
// This file defines the default hardware and UI setup used by the master branch

// Many of the controller UI and default settings can be customized
// by the user by editing the 'localSettings.h' file.

// If a 'localSettings.h' file does not already exist, a new one can be
// created by copying the 'example.localSettings.h' file, and renaming it 'localSettings.h'.
// The 'localSettings.h' file should be included in the .gitignore list and
// will not be overwritten on subsequent controller code updates.
// Edits done direclty in the 'defaultSettings.h' will be overwritten on next pull from the master repo.

// ***** IF USING LOCAL SETTING CUSTOMIZATION *************
// *****************************************************
// Uncomment the following include after a localSettings.h file has been created.
// #include "localSettings.h"



// --------- DO NOT EDIT VALUES BELOW THIS LINE -----------
// --------------------------------------------------------
// Make customizations in the 'localSettings.h' file instead.


// *********************************************************
// *** default #defines in case not Set by localSettings ***
// *********************************************************


// Number of physical lanes available for the system (max of 4 allowed)
// Fewer lanes will support a higher number of DEFAULT_MAX_STORED_LAPS
// Lane count also defines the number of LED racer displays wired
// If using a MAX7219 based LED startlight it will be assume to be device LANE_COUNT +1 (aka at index = LANE_COUNT)
#if !defined ( LANE_COUNT )
  #define LANE_COUNT 2
#endif

// set the maximum number of fastest laps stored during a race
// if this gets too large the system will run out of memory and crash
// recommended max is 20 if LANE_COUNT = 2, 15 if LANE_COUNT = 3, and 10 if LANE_COUNT = 4
#if !defined ( DEFAULT_MAX_STORED_LAPS )
  #define DEFAULT_MAX_STORED_LAPS 20
#endif

// Drag Race lap triggers
// If using only 1 set of finish line triggers for drag racing, set to 'true'
// If using a start and finish line trigger setup for drag racing, set to 'false'
#if !defined ( SINGLE_DRAG_TRIGGER )
  #define SINGLE_DRAG_TRIGGER false
#endif

// If true, this flag tells the controller to do a pre-start countdown when restarting from a pause.
#if !defined ( CTDWN_ON_RESTART )
  #define CTDWN_ON_RESTART false
#endif


// set debounce time in ms
#if !defined ( DEBOUNCE )
  #define DEBOUNCE 500
#endif

// Drag race timeout in seconds, max 255
#if !defined ( DRAG_HEAT_TIMEOUT )
  #define DRAG_HEAT_TIMEOUT 10
#endif


// set the default number of laps used for a standard race
#if !defined ( DEFAULT_LAPS )
  #define DEFAULT_LAPS 10
#endif
// set the default min and sec used for a timed race on bootup
#if !defined ( DEFAULT_SET_MIN )
  #define DEFAULT_SET_MIN 0
#endif
#if !defined ( DEFAULT_SET_SEC )
  #define DEFAULT_SET_SEC 30
#endif
// set the default value of the pre-start countdown
#if !defined ( DEFAULT_COUNTDOWN )
  #define DEFAULT_COUNTDOWN 5
#endif
// set the default number of ticks (ms) between display updates
#if !defined ( DEFAULT_REFRESH_TICKS )
  #define DEFAULT_REFRESH_TICKS 100
#endif

// Length of flash period in ms (time just completed lap is displayed to LED)
#if !defined ( DEFAULT_FLASH_PERIOD_LENGTH )
  #define DEFAULT_FLASH_PERIOD_LENGTH 1500
#endif

// Length of time the start light stays lit after race start for circuit races.
#if !defined ( START_LIGHT_OFF_DELAY )
  #define START_LIGHT_OFF_DELAY 2000
#endif

// Setting Drag Race pre-start countdown range
// Minimum time (seconds) for Drag Pre-Start countdown (Do not set below 2sec)
#if !defined ( DRAG_PRESTART_CNTDWN_BASE )
  #define DRAG_PRESTART_CNTDWN_BASE 4
#endif
// Range of random time added to drag race pre-start countdown.
// Calculated by taking a modulo of the current ms clock time:
// millis() % DRAG_PRESTART_RNDM
// The result will be a whole number between 0 and DRAG_PRESTART_RNDM
#if !defined ( DRAG_PRESTART_RNDM )
  #define DRAG_PRESTART_RNDM 3
#endif

// Lane/Racer's associated with which pin and interrupt byte mask pairs
// !!!! ALWAYS define 4 lanes, regardless of 'LANE_COUNT'.
#if !defined ( LANE1 )
  #define LANE1 {PIN_A0, 0b00000001}
#endif
#if !defined ( LANE2 )
  #define LANE2 {PIN_A1, 0b00000010}
#endif
#if !defined ( LANE3 )
  #define LANE3 {PIN_A2, 0b00000100}
#endif
#if !defined ( LANE4 )
  #define LANE4 {PIN_A3, 0b00001000}
#endif

// Interrupt Hardware (ONLY CHANGE if using a non-Nano based Arduino)
// For ATmega328 based Arduino use 'PCINT1_vect', case sensitive, no quotes
// For ATmega2560 based Arduino use 'PCINT2_vect', case sensitive, no quotes
#if !defined ( PCINT_VECT )
  #define PCINT_VECT PCINT1_vect
#endif
// For ATmega328 based Arduino use 'PINC', case sensitive, no quotes
// For ATmega2560 based Arduino use 'PINK', case sensitive, no quotes
#if !defined ( INTERRUPT_PORT )
  #define INTERRUPT_PORT PINC
#endif


// I2C Address for Adafruit LED Bargraph
#if !defined ( BARGRAPH_I2C_ADDRESS )
  #define BARGRAPH_I2C_ADDRESS 0x70
#endif


// Pin used for the pause button
#if !defined ( PAUSEPIN )
  #define PAUSEPIN PIN_A6
#endif
// Pin used for drag starts
#if !defined ( STARTPIN )
  #define STARTPIN PIN_A7
#endif
// Pin used for the auido buzzer
#if !defined ( BUZZPIN )
  #define BUZZPIN 13
#endif

// *** GAME SOUNDS ***
// BEEP - basic game tone used for button press feedback and lap completion
#if !defined ( BEEP_FREQ )
  #define BEEP_FREQ 4000  // freq in Hz
#endif
#if !defined ( BEEP_DUR )
  #define BEEP_DUR 200    // duration in ms
#endif
// BOOP - secondary game tone used for menu back and start/re-start lap trigger
#if !defined ( BOOP_FREQ )
  #define BOOP_FREQ 1000  // freq in Hz
#endif
#if !defined ( BOOP_DUR )
  #define BOOP_DUR 200    // duration in ms
#endif
// BlEEP - longer tone used for race start and restart after pause.
#if !defined ( BLEEP_FREQ )
  #define BLEEP_FREQ 2000  // freq in Hz
#endif
#if !defined ( BLEEP_DUR )
  #define BLEEP_DUR 600    // duration in ms
#endif

// Audio Mode Default on bootup
// use only 'AllOn', 'GameOnly', or 'Mute', case sensitive, no quotes
#if !defined ( DEFAULT_AUDIO_MODE )
  #define DEFAULT_AUDIO_MODE AllOn
#endif


// LCD Definition of size.
#if !defined( LCD_COLS)
  #define LCD_COLS 20
#endif
#if !defined( LCD_ROWS)
  #define LCD_ROWS 4
#endif
// Location, column index, that sets where to print race and prestart timer clock on the screen
#if !defined( RACE_CLK_POS)
  #define RACE_CLK_POS 8
#endif
#if !defined( PRESTART_CLK_POS)
  #define PRESTART_CLK_POS 11
#endif


// 7-Segment LED bar pinout
#if !defined ( PIN_TO_LED_DIN )
  #define PIN_TO_LED_DIN 2
#endif
#if !defined ( PIN_TO_LED_CS )
  #define PIN_TO_LED_CS 3
#endif
#if !defined ( PIN_TO_LED_CLK )
  #define PIN_TO_LED_CLK 4
#endif



// Main Menu Screen options labels
// max: 20ch
#if !defined( A_SELECT_RACER)
  #define A_SELECT_RACER "A| Select Racers"
#endif
#if !defined( B_CHANGE_SETTINGS)
  #define B_CHANGE_SETTINGS "B| Change Settings"
#endif
#if !defined( C_START_RACE)
  #define C_START_RACE "C| Start a Race"
#endif
#if !defined( D_SEE_RESULTS )
  #define D_SEE_RESULTS "D| See Results"
#endif


// Select Racer's Menu Text
// max: 8ch
// The racer name will be printed to right of this label on same row starting with the 10th character position (column index 9)
#if !defined( A_RACER1 )
  #define A_RACER1 "A|Racer1"
#endif
#if !defined( B_RACER2 )
  #define B_RACER2 "B|Racer2"
#endif
#if !defined( C_RACER3 )
  #define C_RACER3 "C|Racer3"
#endif
#if !defined( D_RACER4 )
  #define D_RACER4 "D|Racer4"
#endif


// Change Settings Menu Text
// max: 11ch
#if !defined( A_SETTING_AUDIO )
  #define A_SETTING_AUDIO " A |Audio"
#endif
// Set the end position of where the race time should be printed.
// Should be chosen to align time digits with 'B_SETTING_TIME' label text.
#if !defined( TIME_SETTING_POS)
  #define TIME_SETTING_POS 14
#endif
#if !defined( B_SETTING_TIME)
  #define B_SETTING_TIME " B |Time       :"
#endif
#if !defined( C_SETTING_LAPS )
  #define C_SETTING_LAPS " C |Laps"
#endif
#if !defined( D_SETTING_LANES )
  #define D_SETTING_LANES "0-4|Lanes"
#endif


// START A RACE MENU
// max: 11ch
// Set the end position of where the lap number should be printed.
// Should be chosen to align number with A_START_RACE_STANDARD text.
#if !defined( START_RACE_LAPS_ENDPOS_IDX )
  #define START_RACE_LAPS_ENDPOS_IDX 19
#endif
#if !defined( A_START_RACE_STANDARD)
  // #define A_START_RACE_STANDARD "A|First to     Laps"
  #define A_START_RACE_STANDARD "A|Laps Race to:"
#endif
// Set the end position of where the race time is printed to the screen
// Should be chosen to align time with label text.
#if !defined( START_RACE_TIME_ENDPOS_IDX )
  #define START_RACE_TIME_ENDPOS_IDX 19
#endif
// Remember to place a ':' at START_RACE_TIME_ENDPOS_IDX - 2
#if !defined( B_START_RACE_IMED)
  // #define B_START_RACE_IMED "B|Most Laps in   :"
  #define B_START_RACE_IMED "B|Timed Race:"
#endif

#if !defined( START_RACE_3RD_ROW )
  #define START_RACE_3RD_ROW "C|Drag Race"
#endif
// Set the end position of the countdown number
// Should be chosen to align countdown value with D_START_RACE_COUNTDOWN text.
#if !defined( START_RACE_CNTDWN_ENDPOS_IDX )
  #define START_RACE_CNTDWN_ENDPOS_IDX 19
#endif
#if !defined( D_START_RACE_COUNTDOWN )
  // #define D_START_RACE_COUNTDOWN "D|Countdown:    Sec"
  #define D_START_RACE_COUNTDOWN "D|PreStart (Sec):"
#endif


// Text to display on main LCD during Pre-Start countdown
#if !defined( TEXT_PRESTART )
  #define TEXT_PRESTART "Your Race Starts in:"
#endif
// First character position of Pre-Start text string
#if !defined( TEXT_PRESTART_STRPOS )
  #define TEXT_PRESTART_STRPOS 0
#endif


// Finishing Place Labels
// MUST be 3 character string
#if !defined( FINISH_DNF)
  #define FINISH_DNF "DNF"
#endif
#if !defined( FINISH_1ST)
  #define FINISH_1ST "1st"
#endif
#if !defined( FINISH_2ND )
  #define FINISH_2ND "2nd"
#endif
#if !defined( FINISH_3RD )
  #define FINISH_3RD "3rd"
#endif
#if !defined( FINISH_4TH )
  #define FINISH_4TH "4th"
#endif


// Results Menu Text
#if !defined( NO_RACE_DATA )
  #define NO_RACE_DATA "-NO RACE DATA-"
#endif
// text displayed while processing race data
#if !defined( COMPILING )
  #define COMPILING "-Compiling-"
#endif
// text displayed on top row of Top Results page
#if !defined( RESULTS_TOP_LBL )
  #define RESULTS_TOP_LBL "C| TOP RESULTS"
#endif

// text used to label 'Best' lap, lap# on the live and FINAL leader board
#if !defined( RESULTS_TOP_TEXT_BEST )
  #define RESULTS_TOP_TEXT_BEST "Best"
#endif
// text used to label 'Best' lap, lap# on the live and FINAL leader board
#if !defined( RESULTS_TOP_TEXT_LAP )
  #define RESULTS_TOP_TEXT_LAP "Lap"
#endif

// Blink rate between Racer#ID and Racer Name on individual results screen.
#if !defined( RESULTS_RACER_BLINK )
  #define RESULTS_RACER_BLINK 2000
#endif

// text displayed on top row of individual racer results menu
#if !defined( RESULTS_RACER_LBL )
  #define RESULTS_RACER_LBL "C| RACER "
#endif
// position to print the racer # following/preceding RESULT_RACER_LBL text
#if !defined( RESULTS_RACER_NUM_POS )
  #define RESULTS_RACER_NUM_POS 10
#endif
// text displayed on top row of the Finish results menu
#if !defined( RESULTS_FINISH_LBL )
  #define RESULTS_FINISH_LBL "C| FINISH"
#endif
// Label displayed above racer's total race time on individual racer's results screen
// max 6 characters
#if !defined( RESULTS_TOTAL_LBL )
  #define RESULTS_TOTAL_LBL "Total"
#endif
// 2nd line of label above racer's total race time.
// max 6 characters
#if !defined( RESULTS_TOTAL_LBLA )
  #define RESULTS_TOTAL_LBLA "Time"
#endif
#if !defined( RESULTS_TOTAL_LBLB )
  #define RESULTS_TOTAL_LBLB "Laps"
#endif


// Start text used on LEDs
// Should not exceed 5ch
#if !defined( TEXT_START )
  #define TEXT_START "Start"
#endif

// 'Pause' text displayed on LEDs
// Should not exceed 5ch
#if !defined( TEXT_PAUSE )
  #define TEXT_PAUSE "PAUSE"
#endif

// text used for '-OFF-' state of audio and enabled lanes/racers
// Code assumes this term is always 5ch long, use spaces to fill if necessary.
#if !defined( TEXT_OFF )
  #define TEXT_OFF "-Off-"
#endif


// RACER_NAMES_LIST defins the list of names available to select from through UI.
// If SONG_BY_PLACE is 'false', the number of terms should match the # of terms in 'RACER_SONGS_LIST"
// !!!! The first term should always be 'TEXT_OFF', it is used to represent an inactive lane
#if !defined ( RACER_NAMES_LIST )
  #define RACER_NAMES_LIST {TEXT_OFF, "Lucien", "Zoe", "Elise", "John", "Angie", "Uncle 1", "Rat2020_longer", "The OG", "5318008"}
#endif
// If SONG_BY_PLACE is 'false', songs are associated with racer name of matching array index
// songs to choose from are defined in the RTTL_songs.h file
// If SONG_BY_PLACE is 'false', the number of terms should match the # of terms in 'RACER_NAMES_LIST"
// !!!! The first term is for the 'off' condition and should reference 'disabledTone'.
#if !defined ( RACER_SONGS_LIST )
  #define RACER_SONGS_LIST {disabledTone, starWarsImperialMarch, takeOnMeMB, airWolfTheme, tmnt1, gameOfThrones, galaga, outrun, starWarsEnd, spyHunter}
#endif
// Song Association
// To associate a song with a racer name, set SONGS_BY_PLACE to 'false'.
// Otherwise, to make songs played related to finishing place, set to 'true'.
// If SONGS_BY_PLACE is 'true', 1st finisher will play RACERS_SONG_LIST[1], 2nd place finisher RACERS_SONG_LIST[2], etc.
// RACER_NAMES_LIST & RACERS_SONG_LIST size do not need to match if SONGS_BY_PLACE is 'true'.
#if !defined ( SONGS_BY_PLACE )
  #define SONGS_BY_PLACE false
#endif


